### PR TITLE
tools/c7n_mailer - lazily import processor modules

### LINKED ISSUE #7853: Deployment of c7n-mailer Fails Due to Missing Package
### Describe the bug

When attempting to deploy the mailer in AWS, it failed with the following error:

`Traceback (most recent call last):
  File "/tmp/custodian/bin/c7n-mailer", line 5, in <module>
    from c7n_mailer.cli import main
  File "/tmp/custodian/lib/python3.9/site-packages/c7n_mailer/cli.py", line 14, in <module>
    from c7n_mailer.gcp_mailer.gcp_queue_processor import MailerGcpQueueProcessor
  File "/tmp/custodian/lib/python3.9/site-packages/c7n_mailer/gcp_mailer/gcp_queue_processor.py", line 17, in <module>
    raise Exception("Using GCP Pub/Sub with c7n_mailer requires package c7n_gcp to be installed.")
Exception: Using GCP Pub/Sub with c7n_mailer requires package c7n_gcp to be installed.`

### What did you expect to happen?

The mailer lambda would deploy in AWS without having to install c7n-gcp.

### Cloud Provider

Amazon Web Services (AWS)

### Cloud Custodian version and dependency information

```shell
Custodian:   0.9.19
Python:      3.9.10 (main, Jan 15 2022, 11:48:04) 
             [Clang 13.0.0 (clang-1300.0.29.3)]
Platform:    posix.uname_result(sysname='Darwin', nodename='PL1USCLT001MAC.local', release='21.6.0', version='Darwin Kernel Version 21.6.0: Mon Aug 22 20:17:10 PDT 2022; root:xnu-8020.140.49~2/RELEASE_X86_64', machine='x86_64')
Using venv:  True
Docker: False
Installed: 

PyYAML==6.0
Pygments==2.13.0
argcomplete==2.0.0
attrs==22.1.0
aws-xray-sdk==2.10.0
bleach==5.0.1
boto3==1.24.87
botocore==1.27.87
c7n==0.9.19
cachetools==5.2.0
certifi==2022.9.24
charset-normalizer==2.1.1
click==8.1.3
colorama==0.4.5
coverage==6.5.0
docutils==0.17.1
execnet==1.9.0
flake8==3.9.2
freezegun==1.2.2
google-api-core==2.10.1
google-api-python-client==2.64.0
google-auth==2.12.0
google-auth-httplib2==0.1.0
google-cloud-appengine-logging==1.1.5
google-cloud-audit-log==0.2.4
google-cloud-core==2.3.2
google-cloud-logging==3.2.4
google-cloud-monitoring==2.11.2
google-cloud-storage==1.44.0
google-crc32c==1.5.0
google-resumable-media==2.4.0
googleapis-common-protos==1.56.4
grpc-google-iam-v1==0.12.4
grpcio==1.49.1
grpcio-status==1.49.1
httplib2==0.20.4
idna==3.4
importlib-metadata==4.13.0
importlib-resources==5.9.0
iniconfig==1.1.1
jaraco.classes==3.2.3
jmespath==1.0.1
jsonpatch==1.32
jsonpointer==2.3
jsonschema==4.16.0
keyring==23.9.3
mccabe==0.6.1
mock==4.0.3
more-itertools==8.14.0
multidict==6.0.2
packaging==21.3
pkginfo==1.8.3
pkgutil-resolve-name==1.3.10
placebo==0.9.0
pluggy==1.0.0
portalocker==2.5.1
proto-plus==1.22.1
protobuf==4.21.7
psutil==5.9.2
py==1.11.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycodestyle==2.7.0
pyflakes==2.3.1
pygments==2.13.0
pyparsing==3.0.9
pyrsistent==0.18.1
pytest==7.1.3
pytest-cov==3.0.0
pytest-forked==1.4.0
pytest-recording==0.12.1
pytest-sugar==0.9.5
pytest-terraform==0.6.4
pytest-xdist==2.5.0
python-dateutil==2.8.2
pyyaml==6.0
ratelimiter==1.2.0.post0
readme-renderer==37.2
requests==2.28.1
requests-toolbelt==0.9.1
retrying==1.3.3
rfc3986==2.0.0
rsa==4.9
s3transfer==0.6.0
six==1.16.0
tabulate==0.8.10
termcolor==2.0.1
tomli==2.0.1
tqdm==4.64.1
twine==3.8.0
typing-extensions==4.3.0
uritemplate==4.1.1
urllib3==1.26.12
vcrpy==4.2.1
webencodings==0.5.1
wrapt==1.14.1
yarl==1.8.1
zipp==3.8.1
```

### Policy

_No response_

### Relevant log/traceback output

_No response_

### Extra information or context

_No response_

### PR DESCRIPTION
**Pull request by ajkerrigan**
The idea here is to help avoid hard dependencies on specific provider packages, notably `c7n-gcp` since we added GCP support for mailer.

Closes #7853

### CODE REVIEW

**Comment by thisisshi** [comment]
we should also add the packages as extras

**Comment by ajkerrigan** [comment]
> we should also add the packages as extras

Good call, meant to do that!

**Comment by thisisshi** [comment]
we should get this in before we release 0.9.20

**Comment by ajkerrigan** [comment]
> we should get this in before we release 0.9.20

Makes sense :+1: . I'm going to have to get a test or two in to get diff coverage up, but first... split out extras for azure and gcp. Let me know if anything seems fishy about it, but it looked pretty clear that the google secrets bit was gcp-specific. I was less confident about any changes beyond that.

File: tools/c7n_mailer/c7n_mailer/utils.py [hunk 1]
```diff
@@ -25,6 +26,35 @@ class Providers:
...
+    """Find the appropriate queue processor for a given provider
+
+    Import provider-specific processor modules lazily. This way running or provisioning
+    the mailer only requires additional provider packages if they're used in the
+    mailer config.
+    """
+    provider = get_provider(mailer_config)
+
+    if provider == Providers.Azure:
+        from c7n_mailer.azure_mailer.azure_queue_processor import MailerAzureQueueProcessor
```
Comment (thisisshi) [inline tools/c7n_mailer/c7n_mailer/utils.py:45]: we should wrap the gcp/azure imports in try/except with instructions on how to install with extras

Comment (thisisshi) [reply tools/c7n_mailer/c7n_mailer/utils.py:45]:
```suggestion
        try:
          from c7n_mailer.azure_mailer.azure_queue_processor import MailerAzureQueueProcessor
        except ImportError:
          raise Exception("Missing c7n-azure package, install via pip install c7n-mailer --extras azure')
```

File: tools/c7n_mailer/pyproject.toml [hunk 2]
```diff
@@ -32,15 +32,23 @@ redis = "^3.4.1"
 jsonpointer = "^2.0"
 jsonpatch = "^1.25"
 types-six = "^1.16.10"
-google-cloud-secret-manager = "^2.8.0"
 importlib-metadata = ">=4.11.1, <5.0"
 
+# Optional packages for additional provider support
+c7n-gcp = { version = "*", optional = true }
+google-cloud-secret-manager = { version = "^2.8.0", optional = true }
+c7n-azure = { version = "*", optional = true }
```
Comment (thisisshi) [inline tools/c7n_mailer/pyproject.toml:40]: i _think_ sendgrid can also be optional?

File: tools/c7n_mailer/c7n_mailer/utils.py (hunk 1 above)
Comment (ajkerrigan) [reply tools/c7n_mailer/c7n_mailer/utils.py:45]: Good call. GCP already had a message (though it was deeper where we try to import from `c7n_gcp`). I added similar behavior for Azure, and included extras recommendations in the failure message.

File: tools/c7n_mailer/pyproject.toml (hunk 2 above)
Comment (ajkerrigan) [reply tools/c7n_mailer/pyproject.toml:40]: Looks like it was initially that way, but #5434 made it work with AWS too.

There may be other deps we can extract over time though :+1: .

**Comment by kapilt** [comment]
structurally looks good to me, but would be good to get some functional verification across at least two providers.

**Review by kapilt — APPROVED** [review]
lgtm

**Pull request merged**